qiskit/pulse/ir/ir.py

```diff
@@ -20,279 +20,107 @@
 """
 
 from __future__ import annotations
-from typing import List
-from abc import ABC, abstractmethod
 
-import numpy as np
+import rustworkx as rx
+from rustworkx import PyDAG
 
-from qiskit.pulse.exceptions import PulseError
+from qiskit.pulse.ir.alignments import Alignment
+from qiskit.pulse import Instruction
 
-from qiskit.pulse.transforms import AlignmentKind
-from qiskit.pulse.instructions import Instruction
+InNode = object()
+OutNode = object()
 
 
-class IrElement(ABC):
-    """Base class for Pulse IR elements"""
-
-    @property
-    @abstractmethod
-    def initial_time(self) -> int | None:
-        """Return the initial time of the element"""
-        pass
-
-    @property
-    @abstractmethod
-    def duration(self) -> int | None:
-        """Return the duration of the element"""
-        pass
-
-    @abstractmethod
-    def shift_initial_time(self, value: int):
-        """Shift ``initial_time``
-
-        Shifts ``initial_time`` to ``initial_time+value``.
-
-        Args:
-            value: The integer value by which ``initial_time`` is to be shifted.
-        """
-        pass
-
-    @property
-    def final_time(self) -> int | None:
-        """Return the final time of the element"""
-        try:
-            return self.initial_time + self.duration
-        except TypeError:
-            return None
-
-
-class IrInstruction(IrElement):
-    """Pulse IR Instruction
+class IrBlock:
+    """IR representation of instruction sequences
 
-    A Pulse IR instruction represents a ``ScheduleBlock`` instruction, with the addition of
-    an ``initial_time`` property.
+    ``IrBlock`` is the backbone of the intermediate representation used in the Qiskit Pulse compiler.
+    A pulse program is represented as a single ``IrBlock`` object, with elements
+    which include ``IrInstruction`` objects and other nested ``IrBlock`` objects.
     """
 
-    def __init__(self, instruction: Instruction, initial_time: int | None = None):
-        """Pulse IR instructions
+    def __init__(self, alignment: Alignment):
+        self._alignment = alignment
 
-        Args:
-            instruction: the Pulse `Instruction` represented by this IR instruction.
-            initial_time (Optional): Starting time of the instruction. Defaults to ``None``
-        """
-        self._instruction = instruction
-        if initial_time is None:
-            self._initial_time = None
-        else:
-            self.initial_time = initial_time
+        self.time_offset = 0
+        self._time_table = {}
```

Comment:
```suggestion
        self._time_table = defaultdict(lambda: None)
```
Maybe? Then you don't need to write `self._time_table.get(index, None)` everywhere.

Reply:
Yeah I was thinking the same thing.